// angular-chat-widget-rasa/esm5/lib/chatbot-rasa.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { ChatAvatarComponent } from './chat-avatar/chat-avatar.component';
import { ChatWidgetComponent } from './chat-widget/chat-widget.component';
import { ChatInputComponent } from './chat-input/chat-input.component';
import { ChatConfigComponent } from './chat-config/chat-config.component';
import { ChatService } from './chatbot-rasa.service';
var ChatbotRasaModule = /** @class */ (function () {
    function ChatbotRasaModule() {
    }
    ChatbotRasaModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, BrowserAnimationsModule],
                    declarations: [ChatAvatarComponent, ChatWidgetComponent, ChatInputComponent, ChatConfigComponent],
                    exports: [ChatWidgetComponent, ChatConfigComponent],
                    entryComponents: [ChatWidgetComponent, ChatConfigComponent],
                    providers: [ChatService]
                },] }
    ];
    return ChatbotRasaModule;
}());
export { ChatbotRasaModule };
