// angular-chat-widget-rasa/esm5/lib/chatbot-rasa.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import io from 'socket.io-client';
import { Observable } from 'rxjs/Observable';
import * as i0 from "@angular/core";
var ChatService = /** @class */ (function () {
    function ChatService() {
        var _this = this;
        this.getMessages = (/**
         * @return {?}
         */
        function () {
            return Observable.create((/**
             * @param {?} observer
             * @return {?}
             */
            function (observer) {
                _this.socket.on('bot_uttered', (/**
                 * @param {?} message
                 * @return {?}
                 */
                function (message) {
                    observer.next(message);
                }));
            }));
        });
    }
    /**
     * @param {?} url
     * @return {?}
     */
    ChatService.prototype.connect = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        var _this = this;
        this.socket = io(url);
        this.socket.on('connect', (/**
         * @return {?}
         */
        function () {
            _this.socket.emit('session_request', { 'session_id': _this.socket.id });
        }));
        this.socket.on('session_confirm', (/**
         * @param {?} remote_id
         * @return {?}
         */
        function (remote_id) {
        }));
        this.socket.on('connect_error', (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.log(error);
        }));
        this.socket.on('error', (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.log(error);
        }));
        this.socket.on('disconnect', (/**
         * @param {?} reason
         * @return {?}
         */
        function (reason) {
            console.log(reason);
        }));
    };
    /**
     * @param {?} message
     * @return {?}
     */
    ChatService.prototype.sendMessage = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this.socket.emit('user_uttered', { 'message': message });
    };
    ChatService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ChatService.ctorParameters = function () { return []; };
    /** @nocollapse */ ChatService.ngInjectableDef = i0.defineInjectable({ factory: function ChatService_Factory() { return new ChatService(); }, token: ChatService, providedIn: "root" });
    return ChatService;
}());
export { ChatService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChatService.prototype.socket;
    /** @type {?} */
    ChatService.prototype.getMessages;
}
