// angular-chat-widget-rasa/esm2015/lib/chat-config/chat-config.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class ChatConfigComponent {
    constructor() {
        this.text = 'Select theme';
        this.themeChange = new EventEmitter();
        this.themes = ['blue', 'grey', 'red'];
    }
    /**
     * @param {?} theme
     * @return {?}
     */
    setTheme(theme) {
        this.theme = theme;
        this.themeChange.emit(this.theme);
    }
}
ChatConfigComponent.decorators = [
    { type: Component, args: [{
                selector: 'chat-config',
                template: `
    <div class="chat-config">
      {{text}}
      <button *ngFor="let item of themes"
              class="btn" [class.btn-active]="item === theme"
              (click)="setTheme(item)">
        {{item}}
      </button>
    </div>
  `,
                styles: [`
    .chat-config {
      padding: 20px;
    }
    .btn {
      padding: 5px;
      margin: 0px 2px;
      border: 1px solid #dedede;
      outline: none;
    }
    .btn-active {
      border: 1px solid #a0a0a0;
    }
    .btn:focus {
      border: 1px solid #333;
    }
  `]
            }] }
];
ChatConfigComponent.propDecorators = {
    theme: [{ type: Input }],
    text: [{ type: Input }],
    themeChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ChatConfigComponent.prototype.theme;
    /** @type {?} */
    ChatConfigComponent.prototype.text;
    /** @type {?} */
    ChatConfigComponent.prototype.themeChange;
    /** @type {?} */
    ChatConfigComponent.prototype.themes;
}
