// angular-chat-widget-rasa/esm5/lib/chat-input/chat-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild, ViewEncapsulation } from '@angular/core';
var ChatInputComponent = /** @class */ (function () {
    function ChatInputComponent() {
        this.buttonText = '↩︎';
        this.focus = new EventEmitter();
        this.send = new EventEmitter();
        this.dismiss = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ChatInputComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.focus.subscribe((/**
         * @return {?}
         */
        function () { return _this.focusMessage(); }));
    };
    /**
     * @return {?}
     */
    ChatInputComponent.prototype.focusMessage = /**
     * @return {?}
     */
    function () {
        this.message.nativeElement.focus();
    };
    /**
     * @return {?}
     */
    ChatInputComponent.prototype.getMessage = /**
     * @return {?}
     */
    function () {
        return this.message.nativeElement.value;
    };
    /**
     * @return {?}
     */
    ChatInputComponent.prototype.clearMessage = /**
     * @return {?}
     */
    function () {
        this.message.nativeElement.value = '';
    };
    /**
     * @return {?}
     */
    ChatInputComponent.prototype.onSubmit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var message = this.getMessage();
        if (message.trim() === '') {
            return;
        }
        this.send.emit({ message: message });
        this.clearMessage();
        this.focusMessage();
    };
    ChatInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'chat-input',
                    template: "\n    <textarea type=\"text\" class=\"chat-input-text\" placeholder=\"Type message...\"\n              #message (keydown.enter)=\"onSubmit()\" (keyup.enter)=\"message.value = ''\" (keyup.escape)=\"dismiss.emit()\"></textarea>\n    <button type=\"submit\" class=\"chat-input-submit\" (click)=\"onSubmit()\">\n      {{buttonText}}\n    </button>\n  ",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".chat-input-text{margin:14px 0 0 14px;height:25px;width:70%;border:0;resize:none;border:none;overflow:auto;outline:0;box-shadow:none;font-size:14px;background-color:inherit;color:inherit}.chat-input-text::-webkit-input-placeholder{color:inherit}.chat-input-text::-moz-placeholder{color:inherit}.chat-input-text::-ms-input-placeholder{color:inherit}.chat-input-submit{margin:14px 12px;float:right;background-color:inherit;color:inherit;font-size:24px;border:0;outline:0}"]
                }] }
    ];
    ChatInputComponent.propDecorators = {
        buttonText: [{ type: Input }],
        focus: [{ type: Input }],
        send: [{ type: Output }],
        dismiss: [{ type: Output }],
        message: [{ type: ViewChild, args: ['message',] }]
    };
    return ChatInputComponent;
}());
export { ChatInputComponent };
if (false) {
    /** @type {?} */
    ChatInputComponent.prototype.buttonText;
    /** @type {?} */
    ChatInputComponent.prototype.focus;
    /** @type {?} */
    ChatInputComponent.prototype.send;
    /** @type {?} */
    ChatInputComponent.prototype.dismiss;
    /** @type {?} */
    ChatInputComponent.prototype.message;
}
