// angular-chat-widget-rasa/esm2015/lib/chatbot-rasa.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import io from 'socket.io-client';
import { Observable } from 'rxjs/Observable';
import * as i0 from "@angular/core";
export class ChatService {
    constructor() {
        this.getMessages = (/**
         * @return {?}
         */
        () => {
            return Observable.create((/**
             * @param {?} observer
             * @return {?}
             */
            (observer) => {
                this.socket.on('bot_uttered', (/**
                 * @param {?} message
                 * @return {?}
                 */
                (message) => {
                    observer.next(message);
                }));
            }));
        });
    }
    /**
     * @param {?} url
     * @return {?}
     */
    connect(url) {
        this.socket = io(url);
        this.socket.on('connect', (/**
         * @return {?}
         */
        () => {
            this.socket.emit('session_request', { 'session_id': this.socket.id });
        }));
        this.socket.on('session_confirm', (/**
         * @param {?} remote_id
         * @return {?}
         */
        (remote_id) => {
        }));
        this.socket.on('connect_error', (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            console.log(error);
        }));
        this.socket.on('error', (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            console.log(error);
        }));
        this.socket.on('disconnect', (/**
         * @param {?} reason
         * @return {?}
         */
        (reason) => {
            console.log(reason);
        }));
    }
    /**
     * @param {?} message
     * @return {?}
     */
    sendMessage(message) {
        this.socket.emit('user_uttered', { 'message': message });
    }
}
ChatService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ChatService.ctorParameters = () => [];
/** @nocollapse */ ChatService.ngInjectableDef = i0.defineInjectable({ factory: function ChatService_Factory() { return new ChatService(); }, token: ChatService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChatService.prototype.socket;
    /** @type {?} */
    ChatService.prototype.getMessages;
}
